[RFC 0077] Stale Issues Amendement

blaggacao opened the pull request:
Based on: https://github.com/NixOS/nixpkgs/pull/100460#issuecomment-708093967

[Rendered](https://github.com/blaggacao/rfcs/blob/da-stale-period-amendemend/rfcs/0077-stale-issues-amendment.md)

---

Based on the first reactions and feedback, unfortunately, I have to clarify this:

- Please read carefully, this RFC does **not** [significantly increase the rate of notifications](https://github.com/NixOS/rfcs/pull/77#issuecomment-708528137).
- This RFC does **not** seek to impose anything on anyone (quite a few comments seem to interpret this RFC as an imposition to their customs or freedoms) &mdash; this is over-interpretation and is **not** the case. It seeks [transparency about the facts](https://github.com/NixOS/rfcs/pull/77/files#diff-fbe3cd7737f03e77c4358e752f5104c766f986bdf639a7cd6bb563bfac6b0a8fR26-R28). If an increase in transparency does prompt (some) people to actions, then this is a positive spill-over effect.
- There seems to be an implicit proxy discussion about different labels going on here (notably the `2.status: *` category). This RFC does not address the expressiveness of the current label structure. It assume it needs improvement, but a general overhaul might be something for a different RFC.
- Unfortunately, _authors_ &mdash; the primary target of the stale-bot's reminder &mdash; are likely to be systemically underrepresented in this RFC's discussion. So please read it with their eyes and needs, too. They are a diverse group &mdash; not only core or regular contributors.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-many-people-are-paid-to-work-on-nix-nixpkgs/8307/85

Emantor commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,95 @@
+chances of (prompt) "success".
+
+An issue that hasn't been interacted with for 60 days or a PR for 90 days
+and in the vas majority of cases deserve an aggregate indication of
```
vas → vast

Emantor commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,95 @@
+[motivation]: #motivation
+
+Under the renewed definition of _stale_ under this PR but the currently
+applied time preiods, spectators based on their common understanding 
```
preiods → periods

nh2 commented:
I'm not in favour.

There is **no practical benefit** of marking things as stale at higher rate. The bot's main purpose is to, infrequently, ask humans whether an issue is still relevant, to prevent years-old issues lying around that are resolved or obsolete. Its purpose is not to determine whether the "conversation is stale".

**We don't need a bot to tell us that a "conversation is stale".** That is already observable from the silence and dates on the discussions. It is not _actionable_. The bot's current purpose is actionable and pragmatic. It prompts a human to triage the issue, 
not to get a conversation going.

There is a very concrete **drawback** of increasing the rate: Spamming people with irrelevant messages, increasing overhead. Many of us are subscribed to thousands of issues; you can calculate how many bot emails per day this means. (This should be added to the drawbacks section.) So they better be actionable. It is unreasonable to expect that an issue has resolved itself within the proposed 90 days; that time is shorter than NixOS's 180 days cadence, and lots of upstream software won't even make a point release within that time.

**Data:** [As of now](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%222.status%3A+stale%22) there are 1750 open issues marked as stale, and 450 stale issues were marked as closed. It might make sense to accelerate the stale time if we were running out of stale issues to triage, but that isn't the case. For those nixpkgs contributors that have extra time and energy available, I recommend to triage some of the 1750, or to work on one of the 2500 non-stale issues. For the contributers that are already packed with work, it would be beneficial to not drive up bot notification spam by 2x.

blaggacao commented:
@nh2 Did you take into consideration https://github.com/NixOS/nixpkgs/pull/100462?

blaggacao commented:
> There is a very concrete drawback of increasing the rate: Spamming people with irrelevant messages, increasing overhead. Many of us are subscribed to thousands of issues; you can calculate how many bot emails per day this means. (This should be added to the drawbacks section.) So they better be actionable. It is unreasonable to expect that an issue has resolved itself within the proposed 90 days; that time is shorter than NixOS's 180 days cadence, and lots of upstream software won't even make a point release within that time.

Is that true? How is it that the rate of notifications goes up?

blaggacao commented:
@nh2 Thanks for your additional arguments (in addition to the ones implicitly addressed by the above mentioned PR). Could you please revise your opinion under the arguments of https://github.com/NixOS/rfcs/pull/77/commits/db779b98fd0b5493bd317dd1fc472f1e3d03c609

nh2 commented:
> Did you take into consideration [NixOS/nixpkgs#100462](https://github.com/NixOS/nixpkgs/pull/100462)?

I did see it linked as related in the RFC, and I just replied to that proposed change in https://github.com/NixOS/nixpkgs/pull/100462#issuecomment-708144150, but I don't think it would change one's opinion on the RFC proposed here (which can be understood independent of the current / alternative bot text).

> Could you please revise your opinion under the additional arguments of [db779b9](https://github.com/NixOS/rfcs/commit/db779b98fd0b5493bd317dd1fc472f1e3d03c609)?

I don't follow the newly added conclusions:

```diff
+One might think, that this increases the load of "spam" notifications on
+subscribers. However, the bot ifirst and foremost prompts and helps the author.
+Subscribers, most of the time, do not have the highest stakes in this interaction. 
```

I don't think this is the case. Most people are subscribers to an issue because they suffer from the same issue, and want to be notified when it's fixed. They would likely have filed it themselves, if somebody else hadn't done it yet.

```diff
+the stalebot has not prompted any action on the
+vast majority of interactions. That means, the stalebot is pretty inefective
+(since ignored).
```

No, it doesn't have to mean that.

* It can mean nobody is interested in the issue. Then the bot fulfilled its purpose here of pointing that out, so it's perfectly effecive.
* It can mean that determining whether some issue still exists is nontrival work, and nobody has had the time yet to reconfirm it.
  For example, if the issue is "NixOS freezes every 30th suspend", that can be quite some work to triage.
* It can mean that the people subscribed to the bug ignored the stale bot emails because they do not have time to post on Github all day, and instead hope for somebody with more time to triage the issue properly.

```diff
+The most plausible root cause is that the stale bot promted
+after an inhumanely long period of time in which the interest of the proponent
+might have shifted
```

That seems a lot less likely than any of the 3 reasons I gave above.

---

Let's keep a in mind:

The key thing that moves issues forward is people writing code.

That's how stuff gets done; hard work with significant time investment. Talking and pinging rarely resolve issues.

blaggacao commented:
@nh2 Thanks for your additional revision, just to let you know you might eventually want to update it with the latest commit: https://github.com/NixOS/rfcs/pull/77/commits/887785b7fbeaa0f2ad1d174b8d676fa8a4255721

blaggacao commented:
> I don't think this is the case. Most people are subscribers to an issue because they suffer from the same issue, and want to be notified when it's fixed. They would likely have filed it themselves, if somebody else hadn't done it yet.

If a bot prompts the author, subscribers are not direct addressees of the bot. Can you propose a formulation that makes this clear? (I'll also try, since it seems to be misunderstood).

> It can mean nobody is interested in the issue. Then the bot fulfilled its purpose here of pointing that out, so it's perfectly effective.

That would pretty much mean the bot is _not_ actionable (it didn't trigger action), which is what we are trying to address in here.

> It can mean that determining whether some issue still exists is nontrival work, and nobody has had the time yet to reconfirm it.
For example, if the issue is "NixOS freezes every 30th suspend", that can be quite some work to triage.

_Actionably_ adressed by https://github.com/NixOS/rfcs/pull/77/commits/887785b7fbeaa0f2ad1d174b8d676fa8a4255721 &mdash; you forgive me the pointer.

> It can mean that the people subscribed to the bug ignored the stale bot emails because they do not have time to post on Github all day, and instead hope for somebody with more time to triage the issue properly.

If so, due respect (in my stance) requires those group to ask for it (anew) instead of _boldly_ throwing in an issue and _cowardly_ hoping for the best. Besides the words, I guess you see the point.

blaggacao commented:
@nh2 Judging from https://github.com/NixOS/rfcs/pull/77#issuecomment-708151224, some additional rewording should have clarified the arguments:  https://github.com/NixOS/rfcs/pull/77/commits/f647a3563e57518969679a51b236e494dfed4081

Would you bother to revise again your judgment, since I consider I have duly and substantially addressed all of your concerns / arguments. Please let me know, if I should have missed something. :wink:

suhr commented:
I'm strongly against this proposal.

> People with vested interest should see the stale label and be prompted to think: "Oh, I need this, too. Damn, it's stale. Let's have a look and do something to help out."

Abusing sense of urgency might be fine in marketing, but is *highly inappropriate* for a open source project. Despite that, people who are working on an issue are not obligated to report their status weekly. This is not their daily job after all.

And of course, we don't need a bot to tell us that a "conversation is stale". The bot was introduced to help people to close issues that are *no longer relevant.* And in this regard, ~450 closed issues is not bad at all.

7c6f434c commented:
>If a bot prompts the author, subscribers are not direct addressees of the bot. Can you propose a formulation that makes this clear? (I'll also try, since it seems to be misunderstood).

You get notifications for new comments in threads where you were mentioned, you have commented, your review has been requested, or you have subscribed. The former three are considered «updates in threads you participate».

Ma27 commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,135 @@
+and in the vast majority of cases deserves this hint out of fariness.
+
+Honesty and transparency are by far more effective than any perceived
+political correctness (judging by the discussions on RFC 51).
```
This sentence seems more like a subjective complaint rather than an actual reasoning why you disagree with the results of the mentioned RFC.

blaggacao replied:
It is a claim anyone can disagree. I hope dearly hope though that honesty is higher rated in the nix community than political correctness. Would you disagree?

Ma27 replied:
Yes, everyone can disagree if they want to. However, this is a technical document about a change affecting our ecosystem, so this is IMHO not the place to write down personal impressions as if those would be a widely-accepted truth.

blaggacao replied:
I agree, but it is the motivation section. What's wrong about being motivated based on a personal judgment? But I agree it should not do this:
> write down personal impressions as if those would be a widely-accepted truth.

blaggacao replied:
I reworded in https://github.com/NixOS/rfcs/pull/77/commits/120af202277c13d6bb084adbb614464807496b8b

Hence marking this particular feedback as resolved. Thanks, the document is better now! Feel free to reopen, if there is anything more to it...

Ma27 commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,135 @@
+
+People with vested interest should see the _stale_ label and be prompted
+to think: "Oh, I need this, too. Damn, it's stale. Let's have a look
+and do something to help out."
```
I may be the only one, but if I stumble upon a problem (or want a new feature) I don't really care if an existing issue about it is stale or not, I'll try to provide some help to get it done.

blaggacao replied:
That's nice! But it does it invalidate the implicit argument about nor the benefits about transparency, does it?

Ma27 replied:
I wanted to express that I don't think that a `stale`-label has a benefit for this very case you described here - that one only decides to look at an issue they're affected by just because it's marked as stale now.

blaggacao replied:
You are absolutely right about this. But also imagine a scenario (not so remote):

- Filtering by stale label
- Skimming the titles
- Be prompted by the title something like: "Oh, this sounds interesting, but too bad, it's stale. Maybe it could need a hand?"

Or gaining a quick overview of the _staleness_ of the project, like to make a quick decision on weather to bet on the project or not.

blaggacao replied:
There are many more possible scenarios for illustrating the benefits of transparency. Maybe we could just short cut to agree that it's a good thing we should try to improve?

blaggacao replied:
I reworded in https://github.com/NixOS/rfcs/pull/77/commits/120af202277c13d6bb084adbb614464807496b8b

Hence marking this particular feedback as resolved. Thanks, the document is better now! Feel free to reopen, if there is anything more to it...

blaggacao replied:
If reformulated the motivation to better fir the purpose of this RFC. Thanks for prompting this possibility of misinterpretation. It is much clearer now. &rarr; https://github.com/NixOS/rfcs/pull/77/commits/609184fd539a1777a5250920b343b58b2a43c621

Feel free to un resolve, if there is anything more to it that I should take into consideration.

Ma27 commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,135 @@
+meaning by virtue of this very PR, there is no reason to adversely interpret
+a bot spelling out the facts.
+
+Regular subscriber might get notiviations they do not find immediatly
```
s/notiviations/notifications

Ma27 commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,135 @@
+should have prompted the _author_ into "furthering the cause" (whatever that
+might be in a particular context), judging by the 1750 unattended interventions,
+it blatantly failed. Since any such furthering would have left traces the stalebot
+would have picked up in any way.
```
Expecting people to do something they may not have time/energy/motivation to is IMHO wrong in an open source project where a majority contributes in their *free* time.

blaggacao replied:
The only underlying expectation expressed in this RFC is to clean up after oneselves.

How come, you interpret that way?

blaggacao replied:
If removed this section and reworked the interpretation of the underlying available data. Feel free to un resolve if there is anything more to it that should I take into account. Thanks! for raising the issue about this part and helping me to make it clearer.

Ma27 commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,135 @@
+it blatantly failed. Since any such furthering would have left traces the stalebot
+would have picked up in any way.
+
+One reason for this might be that by the time the stae bot intervenes, the
```
s/stae/stale

blaggacao commented on rfcs/0077-stale-issues-amendment.md:
```diff
@@ -0,0 +1,135 @@
+individuals did choose not to interact on a particular issue or PR
+for an extended period of time._
+
+_180 days is too long of such extended period of time. People do relate
```
> (as in: no "stale") 

https://www.merriam-webster.com/dictionary/stale

> : impaired in vigor or effectiveness

How does vigor and effectiveness relate to "NixOS release cycle + ~1 month"?

blaggacao replied:
I made that clear again in https://github.com/NixOS/rfcs/pull/77/commits/6708f3abfe43f57d39d335030d09525f3feb1f3e

Thanks for prompting me for improving this. Not redefining a common english word with any special (diverted) meaning is really the key to understanding this RFC.

blaggacao replied:
Can we mark this particular misunderstanding resolved?

blaggacao commented:
@suhr May I kindly ask you to lay out your objections in detail (not summarily)? It seems you put forth false pretenses about this RFC.

> Abusing sense of urgency might be fine in marketing

What makes you think this RFC is attempting to do so? It is not, so if you see offending formulations, please make suggestions on how to correct them.

> And of course, we don't need a bot to tell us that a "conversation is stale".

How does this relate to https://github.com/NixOS/rfcs/pull/77/files#diff-fbe3cd7737f03e77c4358e752f5104c766f986bdf639a7cd6bb563bfac6b0a8fR102-R105 ?

blaggacao commented:
Can those people please bother to submit an amendment to the drawback section in order to further productive discussion rather than blatant opinion-poll nay saying? RFCs are about discussion and finding the truth, not about vote!

![image](https://user-images.githubusercontent.com/7548295/96012674-f605c580-0e09-11eb-81e0-4fce54baffb5.png)

blaggacao commented:
@7c6f434c 
> You get notifications for new comments in threads where you were mentioned, you have commented, your review has been requested, or you have subscribed. The former three are considered «updates in threads you participate».

This RFC does not significantly increase the rate of notifications (unless they are actionable, because the bot triggered further action &rarr; hence _valuable_ notification). Why do readers tend to assume this (wrongly)?

7c6f434c commented:
>+_180 days is too long of such extended period of time. People do relate
>
>> (as in: no "stale") 
>
>https://www.merriam-webster.com/dictionary/stale
>
>> : impaired in vigor or effectiveness
>
>How does vigor and effectiveness relate to "NixOS release cycle + ~1 month"?

How does dictionary definition relate to tradeoffs that have already been negotiated in a lot of detail before accepting the previous RFC?

Painfully agreed-upon balance of triage and annoyance beats linguistic prescriptivism.

blaggacao commented:
> How does dictionary definition relate to tradeoffs that have already been negotiated in a lot of detail before accepting the previous RFC?

Because it (legitimately) questions the underlying assumptions. Currently, and based on the reactions, I have an ever more strong feeling that the root causes have not been exhausted in the previous RFCs

> Painfully agreed-upon balance of triage and annoyance beats linguistic prescriptivism.

I agree! I really do. But I think the underlying assumptions are wrong, also the (reformulated) data section provides some strong suggestions about the causes. (which might obviously be wrong - but interpreting facts is never easy. At least I'm being very plausible.)

7c6f434c commented:
>+Honesty and transparency are by far more effective than any perceived
>+political correctness (judging by the discussions on RFC 51).
>
>It is a claim anyone can disagree. I dearly hope though that honesty is higher rated in the nix community than political correctness. Would you disagree?

Look, GitHub already allows you to filter by «last update: <2 months ago». There is no information concealment, no information misrepresentation, and — a rare case for Nix* ecosystem — GitHub is even able to process and act on this information.

So one thing that label does not do is it does not provide new information, and thus does not affect honesty.

It is used as a trigger for a possible triage. So many people here judge it from the point of view of an optimal way of taking some action. Your proposal is, in my and not only my opinion, further from the optimum than the current situation.

7c6f434c commented:
>Can those people please bother to submit an amendment to the drawback section in order to further productive discussion rather than blatant opinion-poll nay saying? RFCs are about discussion and finding the truth, not about vote!

RFCs are about building a community consensus. If at the very beginning there is already a near-consensus against the proposal, it would be useful to find out.

More drawbacks submitted as proposed changes won't actually give new information, discussion has it all.

And, other people who are in favour, please join the discussion and present your concurring voices. It skews the perception of the discussion if you do not.

>![image](https://user-images.githubusercontent.com/7548295/96012674-f605c580-0e09-11eb-81e0-4fce54baffb5.png)

blaggacao commented:
> It is used as a trigger for a possible triage. So many people here judge it from the point of view of an optimal way of taking some action. Your proposal is, in my and not only my opinion, further from the optimum than the current situation.

This is true, but what I'm advocating that it should not take a stance _at all_ (at least avoid it). Taking stances is difficult and always political. Who are we too judge what people might potentially get out of the notion of _stale_ (the merriam webster notion)? We cannot even imagine all positive uses.

On the other hand, I think I sufficiently pointed out that this RFC does not worsen the optimum, or at least opponents have not succeeded in making their argument clear and understandable. I have reworked the wording so to help people to ship around some of those misinterpretations. I would very much like an opponent to make a spot on case and aI would be more than happy to loose an argument. Really!

blaggacao commented:
> More drawbacks submitted as proposed changes won't actually give new information, discussion has it all.

I can agree with that and I'm doing my best to transact those discussions in new commits. To _all_: please be fair and review your  stance with the best of intentions based on those amendments. :+1:

7c6f434c commented:
>@7c6f434c 
>> You get notifications for new comments in threads where you were mentioned, you have commented, your review has been requested, or you have subscribed. The former three are considered «updates in threads you participate».
>
>This RFC does not significantly increase the rate of notifications (unless they are actionable, because the bot triggered further human action &rarr; hence _valuable_ notification). Why do readers tend to assume this (wrongly)?

A non-trivial share of issues get stale-bot request «what's the status», then update from someone «checked, still happens». Your proposal triples this traffic without anything useful actually happenning there.

Also, stale-bot is bursty, so its notifications come in blocks and are annoying. This happens now, and your proposal will increase the bursts. 

And then there is the issue of going through 4 months of backlog in a week, this will be already be annoying enough to make this entire story not worth it, in my opinion.

blaggacao commented:
> A non-trivial share of issues get stale-bot request «what's the status», then update from someone «checked, still happens». Your proposal triples this traffic without anything useful actually happenning there.

Thank you this is interesting and indeed it is a missing chunk of data to consider. I'll try to put together estimates or check if a github filter can enlighten. Especially we have to put this into relation with overall issues / stale + open issues and stale + closed issues to afford an informed judgment.

Please note this RFC is accompanied by an [improvement in the stale bot text](https://github.com/NixOS/nixpkgs/pull/100462). We are probably safe to assume, that this might have a cross-effect with out without this RFC on the stale bot effectiveness.

blaggacao commented:
> So one thing that label does not do is it does not provide new information, and thus does not affect honesty.

If honesty and transparency are adjacent, the increase of transparency has an undeniable indirect effect on honesty. So in that light, a label does increase transparency (at scale) and by extension honesty (at scale).

blaggacao commented:
> And then there is the issue of going through 4 months of backlog in a week, this will be already be annoying enough to make this entire story not worth it, in my opinion.

I do not understand what you are saying here. Can you elaborate?

7c6f434c commented:
>> How does dictionary definition relate to tradeoffs that have already been negotiated in a lot of detail before accepting the previous RFC?
>
>Because it (legitimately) questions the underlying assumptions. Currently, and based on the reactions, I have an ever more strong feeling that the root causes have not been exhausted in the previous RFC.

Root causes of what? Root cause of issue accumulation is that nobody qualified to resolve them has allocated enough effort to resolving them. This RFC marginally makes situation worse, because now there are slightly more notifications to dismiss.

>> Painfully agreed-upon balance of triage and annoyance beats linguistic prescriptivism.
>
>I agree! I really do. But I think the underlying assumptions are wrong, also the (reformulated) data section provides some strong suggestions about the causes. (which might obviously be wrong - but interpreting facts is never easy. At least I'm being very plausible.)

Assumptions are irrelevant. Current state represents negotiation on the tradeoffs of effects. It won't change if the label is renamed, or if we assume some inherent issue priority without bringing together people who prioritise their efforts in the same way.

blaggacao commented:
> Root causes of what?

Root cause of stale issues not being freed up by their owner in an act of "step down" to make the way for others to take up. Root cause of (in some instances) creating issues without assuming the responsibilities that go along with it. Root cause of many issues being "impaired in vigor and effectivenss" to a point their chances are reduced to a minimum. People might want to use a threshold to make life reasonable and moderate.

> It won't change if the label is renamed, or if we assume some inherent issue priority without bringing together people who prioritise their efforts in the same way.

This is a not a goal of this RFC. How can I be clearer? &mdash; once adopted, we might by more options to work towards that goal, but it is not immediately achieved by this RFC.

7c6f434c commented:
>> So one thing that label does not do is it does not provide new information, and thus does not affect honesty.
>
>If honesty and transparency are adjacent, the increase of transparency has an undeniable indirect effect on honesty. So in that light, a label does increase transparency (at scale) and by extension honesty (at scale).

This label kind of provides some minimal amount of extra information compared to what already there, but it is pretty minor and not worth the effort as information. So no, stale-bot is not there to increase transparency, it is to prompt actions, and for that «too often» is also _not_ «better».

7c6f434c commented:
>> And then there is the issue of going through 4 months of backlog in a week
>I do not understand what you are saying here. Can you elaborate?

Immediately after reconfiguration, the bot will leave _a lot_ of comments in a short time.

blaggacao commented:
> This label kind of provides some minimal amount of extra information compared to what already there, but it is pretty minor and not worth the effort as information. So no, stale-bot is not there to increase transparency, it is to prompt actions, and for that «too often» is also _not_ «better».

By extension your reasoning, it is an interesting thought to prevent stale bot from posting and only do the label thing.

7c6f434c commented:
> Root cause of (in some instances) creating issues without assuming the responsibilities that go along with it.

Nix* is a project that survives by accepting that people can provide something of value without committing to a huge obligation. I do not see any inherent obligations of an issue reporter, especially not beyond a reproduction case and answering requests.

Nixpkgs is a database. NixOS is a database. Issue list is a database of observations. Nothing more.

Root cause of our issue tracker being a mess is GitHub not having a real issue tracker.

7c6f434c commented:
> To all: please be fair and review your stance with the best of intentions based on those amendments.

Speaking of transparency and lack of politeness.

By now, I am not sure you have enough fluency in relevant gotchas — of how GitHub works, or how Nix* workflows live in practice, or what are the underlying preferences of different project participants evidenced by the discussion of the previous RFC — to be effective at convincing people as a sole proponent of this change.

This is a separate issue from this change being (obviously to me, but that's subjective) a bad fit for the observed spread of preferences.

blaggacao commented:
@7c6f434c Thanks a lot for your arguments, I'll definitely incorporate quite some of it into the document. If we can tip the document and turn it against itself. I'm very happy and satisfied as it reflects a very fruitful discussion! Thank you! :heart: 

> Nixpkgs is a database. NixOS is a database. **Issue list is a database of observations.** Nothing more.

This is a very respectable stance. And I really agree with:

> Root cause of our issue tracker being a mess is GitHub not having a real issue tracker.

But let me put an analogy from the analogue world: A database is like a library, and a library only fulfills it's purpose (whatever that actually is) if it is curated by some means. Fact is, for the very least, that all libraries I know of, are curated. So how do we curate a database (aka library) of issues, while not having the resources to do so?

An underlying stance of this RFC is to promote responsible self-governance. And the decisive argument is that it is less costly for everybody involved if the memory is still somewhat fresh and they can still afford to switch into the context of an issue within reasonable time.

7c6f434c commented:
> Fact is, for the very least, that all libraries I know of, are curated.

This is, to the first three orders of approximation, false of the most crucial primary libraries. The main library of a large country is often supported by a law making it illegal to sell ten hardcopies of anything without sending a copy to the central library. These copies will be indiscriminately stored for decades.

blaggacao commented:
> By now, I am not sure you have enough fluency in relevant gotchas

I'm a new breed. I don't. Some people agree that's a good influence in diversity and different backgrounds. Especially when summing up this year's NixCon talks in my own words: "How to make nix great after 20 years of shadow?"

blaggacao commented:
> This is, to the first three orders of approximation, false of the most crucial primary libraries. The main library of a large country is often supported by a law making it illegal to sell ten hardcopies of anything without sending a copy to the central library. These copies will be indiscriminately stored for decades.

Nice catch! But you are implicitly supporting the argument of this RFC, no one goes to the central library for every day practical life. So our central library _even_ (at the risk of provoking strong opposition - hehehe :wink:) could be _closed_ issues.

7c6f434c commented:
> Some people agree that's a good influence in diversity and different backgrounds.

I do not say that you carefully arguing for your position is a bad thing! But when a newcomer tries to argue for a change to something that has tired everyone during the first discussion, well, sometimes this fails hard. Just because much more novel and heavy arguments are needed to move the median position.

blaggacao commented:
> novel and heavy arguments are needed to move the median position.

Novelty &rarr; data section interpretation. I think it really qualifies (at least to certain extend).
Weight &rarr; if RFCs where designed to bias against agility and cement conservatism, then I agree. I also agree they are not the best format to move forward and iterate quickly (and fail!). But since the overall community suffers from a low rate of advancement (guix has already bootsrapped gcc, iirc), might we just ease a little on this one for the sake of evolution?

On the other hand, what a mistake to cement those numbers in an RFC in the first place? Where is the necessary room for gradual improvement? But that was other people's call at that time.

7c6f434c commented:
> But you are implicitly supporting the argument of this RFC, no one goes to the central library for every day practical life.

No one goes to library for everyday practical needs, full stop. Curated collections a people's private ones. I would not bet on people _preferring_ their closest library to have _less_ things available, it's a cost issue, not the goal. (Of course, many people don't care what their library has because they mostly use it as an interface to inter-library loans, but that makes them users of meta-libraries, often spread enough to be effectively uncurated)

> Maybe the root of the opposition despite the pointed arguments for a change here still being that people are too dearly attached to their overloaded interpretation of either closed or stale?

What is «too»?

People who have done and continue to do a lot of work for Nix* have debated in finest details what workflows related to issues they prefer to use, and found a compromise that did not go too far from the preferences of the people involved

blaggacao commented:
The library issue is an excellent analogon to reason about this. It circles back to better issue labelling, of which the nix project could need a fair more share through triaging (by a human). Since that is not available, this RFC seeks to improve auto-triage. As if at the entrance of the library you would put somebody and he hands over a book to every visitor saying: Please judge if this should be in the first row, second row, third row, etc. Is it for children or is it scientific? etc.

Applied to this RFC we could transpond: "Here is a book. Have a look, and then come back in 60 days instead of in 180 days and put it in the row." &mdash; bad comparision in most aspects, but one: mental cost of context switching.

> What is «too»?

"Too" is when a third party cannot come to the same conslusion of balancing the tradeoffs even when trying really hard to (theleologically) take into account all the arguments. It means that a third party percieves an outcome to negatively suffer from a bias. Thanks you asked! It could have been interpreted in so many less constructive ways!

7c6f434c commented:
> data section interpretation. I think it really qualifies (at least to certain extend).

You need novel _and_ heavy.  As you noticed, this argument can also prompt a reaction «thanks for collecting the data, works as designed».

> Weight → if RFCs where designed to bias against agility and cement conservatism, then I agree.

RFCs are intended to provide as much agility as possible… to inherently failure-prone process of building consensus of non-objection to some decision.

And you are promoting _overturning_ a hard-earned decision on an issue discussed to exhaustion. Yes, you need an argument heavy enough to plausibly shift the balance of a long negotiation.

And also, by the way, using a dictionary as an argument has _negative_ weight in my opinion. Basically, you are getting the things wrong way. We agreed on some behaviour, and slapped some random label on it. You can use a dictionary to argue that another label is much better. If you use a dictionary to argue from a pretty carelessly chosen label towards changing a pretty painfully negotiated behaviour, well, if it works, we have a larger problem.

blaggacao commented:
> Yes, you need an argument heavy enough to plausibly shift the balance of a long negotiation.

Agreed. So we can basically say:

- This is actually a good idea.
- Or it isn't.
- Anyway, do we want to discuss this _again_?
- Maybe not.
- Let's not burn our fingers.

If this is the prevailing stance (in my words), how can we bring about coordinated change, where a some parts are seemingly unweighty but in combination unfold their power towards greater goals? We cannot bring about changes in small steps, since they might not be heavy enough and we cannot proof the overarching goal is heavy enough since we cannot make the small steps. So how do we escape this evolutionary deadlock? (If I was advocating for it &mdash; which I am).

> And also, by the way, using a dictionary as an argument has negative weight in my opinion. Basically, you are getting the things wrong way. We agreed on some behaviour, and slapped some random label on it. You can use a dictionary to argue that another label is much better. If you use a dictionary to argue from a pretty carelessly chosen label towards changing a pretty painfully negotiated behaviour, well, if it works, we have a larger problem.

I think this is another input for amendments to this RFC. Thank you! But I would add:

> We agreed on some _random_ behaviour,  and slapped some random label on it.

I'm repeating myself, but this RFC argues for a change in behaviour on the bases of a (perceived) novelty. Clearly framing the semantics is a way to ease people into the arguments. Since as you see, most people are opposed because of contextual private (to the nix ecosystem) information (or rather interpretation). This very interpretation has it's own issue. We have a real problem when:

- _stale_ means _invalid_
- _closed_ means _shut up_
- What does `2: status: invalid` mean again? &rarr; _stale_?

You see the point.

7c6f434c commented:
> Please judge if this should be in the first row, second row, third row, etc.

The absolute majority of entries in a work-oriented library are sorted by very very rough subdivision, then alphabetically. I guess if you only have fiction you can afford more extravagant approaches, but I hope no issues in our tracker are pure works of fiction.

>  It means that a third party percieves an outcome to negatively suffer from a bias.

See, trade-offs are trade-offs between preferences of different people. The arguments serve to explain the preferences and be able to negotiate about them. If some change makes my workflow inconvenient or annoying or whatever, it is both subjective (specific to me), but also absolutely impossible to fix by an argument that an alternative workflow I have tried and rejected is better. Everyone is an absolute authority on their preference among things they have already tried. And all of them are _right_, even though they do not agree. It is not bias, it is the actual hard work of negotiating trade-offs between values of different people.

blaggacao commented:
Spot on! Thanks!

Except for:

> workflow I have **tried**

A shorter time to first interaction has never been tried, as far as I know on the nix repositories. So that rises the questions are all arguments about an increase in unwanted notifications hypothetical? (I'm genuinely wondering).

It would be very unsatisfying if that would be an argument against the lack of heaviness of this proposal.

7c6f434c commented:
> So how do we escape this evolutionary deadlock?

Step one: find someone who has significant experience and also things it is a deadlock, and not a stable consenus with active users of whatever is being discussed mostly content. Failure might mean that this «deadlock» serves a purpose and fulfills it.

> I'm repeating myself, but this RFC argues for a change in behaviour on the bases of a (perceived) novelty. 

Nope, no novel arguments you provide are convincing.

> This very interpretation has it's own issue. 

Connotations are a thing. Yes, there are people who prefer «closed» to mean «no action useful». Yes, it is a valid and reasonable preference. No, it is not the only one.

> We agreed on some random behaviour, and slapped some random label on it.

No, you do not understand what was happenning, which is why your proposal will probably fail. 

The _behaviour_ was discussed, negotiated, carefully debated. Then name… well, nobody cared enough to discuss it. This is the crucial difference.

>A shorter time to first interaction has never been tried, as far as I know on the nix repositories.

Well, if stale-bot is the first interaction, then the issue is doomed for the time being. Might make sense to check if it got fixed by some refactoring in the next release, 180 days later.

But stale-bot is not a complete workflow. It is an influence on various workflows people have.

> are all arguments about an increase in unwanted notifications hypothetical

We know what are the results of using stale-bot. Changing a parameter with pretty linear effects will lead to pretty predictable outcomes.

7c6f434c commented:
> And if this is perceived basis enough for opposition, I can swallow it, too. 

No, the _problem_ is that your arguments are not convincing enough to renegotiate that, honestly, pretty long debate.

You being new is just the reason why you do not feel the full extent of this problem.

blaggacao commented:
Why do other communities come to different conclusions? And why do they have actionable responses rates to stale bots superior to our 20%? (Ok I just made that up. But from my interactions to kubernetes, it is at least my perception.) &mdash; Thin grounds. Sorry.

If this claim proved to be true, does it mean that other communities make more efficient use of their resources?

7c6f434c commented:
> Why do other communities come to different conclusions? And why do they have actionable responses rates to stale bots superior to our 20%?

Having a well-defined scope distinct from «package like half the global open source ecosystem and make it work together on a new technical foundation» helps.

Really, Nixpkgs has atypical pattern of relations between parts, and also has a relatively rare problem of «no, it is not a cheap action just to check if everything at least builds», and our domain experts are experts in areas too far apart… A lot of best practices are hard to translate.

>Does it then mean we have a surplus of resources so we can waste them by not fine tuning our workflows to the purported data?

Can we afford wasting the quality issues? Yes, in the sense that our bottlenecks are elsewhere.

blaggacao commented:
To conclude, the project might want to improve it's strategies and tactics to harness the resources (and wisdom) of the crowd in purposeful ways. This is an overarching goal, which in itself is _very_ significant.

This RFC is trying to fine tune workflows towards this goal, but in itself is controversial (thanks to RFC51) and by extension lacks the criterion of isolated significance.

Since things start to move constructively into the direction with the above overarching goal, the best conclusion would be probably to leave this RFC around to allow people to reshape their opinions (or reinforce either way).

Example motions of better harnessing the resources of the multitude:

- The concept of flakes
- Some of this year's nixcon talks
- Adopting nix for dev shell environments (eg `numtide/devshell`) &mdash; ¿Entice the developer!
- ...

7c6f434c commented:
It might be that in a couple of days we find some opinions different in an interesting way (I mean, so far it is roughly ten opposing people, there are way more people interested in tooling)

I would mention NUR, too

blaggacao commented:
@suhr 

> Abusing sense of urgency might be fine in marketing, but is highly inappropriate for a open source project. Also, people who are working on an issue are not obligated to report their status weekly. This is not their daily job after all.

- This RFC does not pretend to usurp sense of urgency. To the contrary it pretends to counteract the evil force of _fading memories and interests_: a highly relevant force in open source projects.

- This RFC does not propose to prompt anybody into reporting their status. It is not obvious how you gained this interpretation? &mdash; for clarity why this is not obvious: this RFC is very specific about the meaning of stale (referring to Merriam Webster). So if somebody does not want to report ones status, then the definition of stale is still valid. Even, if an issue is effectively **imparied in vigor and effectivenss** responding with the goal to un-stale can be considered a bad practice and a fight against an imaginary enemy: a waste of resources by every account. I think this RFC can stress this aspect.

> And of course, we don't need a bot to tell us that a "conversation is stale". The bot was introduced to help people to close issues that are no longer relevant. And in this regard, ~450 closed issues is not bad at all.

- I might need to correct you: the bot was introduced by the following motivation...

  > We have a large number of open issues that have accumulated over the years. Not all of them are still valid and **need our attention**.
  > By marking stale issues, we can more easily filter issues for ones that have **at least one person interested in them**.

- There is no clear notion of _relevance_ (as you claim) to be found within the accepted version of the RFC. The most specific part suffices with the technical definition of a stale issue / PR. Interpretation is left on the spectator.

- 450 is an absolute number, it is of questionable advancement to the discussion when not put it into perspective. It's (only) 20% percent of stale issues. We still have to assess the rate of stale issues which triggered desired action. (since they are currently no longer stale and escape the filter). But a vast majority of 1750 stale and open issues compared to roughly 4.4k open issues does suggest, that a significant portion of stale-bot intervention is not acted upon.

suhr commented:
> To the contrary it pretends to counteract the evil force of fading memories and interests: a highly relevant force in open source projects.

But interests are largely driven by need. And when you *need* something, you usually don't forget about it.

I'm quite skeptical about encouraging people actions with bots. It works for answering “yeah, I don't have this problem anymore”. Beyond that, this is not working, it just annoys people.

> this RFC is very specific about the meaning of stale (referring to Merriam Webster)

I must say, this is a less common definition of “stale”. The more common are:

- tasteless or unpalatable from age
- impaired in legal force or effect by reason of being allowed to rest without timely use, action, or demand

> There is no clear notion of relevance (as you claim) to be found within the accepted version of the RFC.

https://dictionary.cambridge.org/dictionary/english/relevant

- connected with what is happening or being discussed
- **correct or suitable for a particular purpose**

Anyway, let's avoid dictionary debates.

blaggacao commented:
> But interests are largely driven by need. And when you need something, you usually don't forget about it.

Exactly! Henceforth, if you forget about something, and leave it around stale, you just created a _negative marginal external effect_ on somebody else that compunds quite heavily with a database of 4.4k open issues. Some of the arguments expressed against this RFC do actually reflect quite pointedly this compound negative effect.

> I'm quite skeptical about encouraging people actions with bots. It works for answering “yeah, I don't have this problem anymore”. Beyond that, this is not working, it just annoys people.

Well, I think the evidence may speak against this: by my analysis in this RFC it "worked" for roughly 42% of interactions after a loooong period (problem of fading interest). The whole point of this RFC is to lower time to first interaction so that it actually gets _less annoying & more relevant_ (since concurrent with interests).

Apart from only leaving a non incremental comment with the only purpose of un staling an issue, which is bad habit, as we established:

- Closing an issue is a relevant action (authors and maintainers can close) - if I'm not more interested in old issues, I regularily close them, even if they still persist, since I effectively block others from ownership.
- an incremental comment at a crucial moment (not 180 days later) but when the context is still somewhat in memory, in my experience, a fair share of times triggers a chain of actions that ultimately result in resolution.

blaggacao commented:
As a first collateral of this motion, there now is:

https://github.com/NixOS/nixpkgs/pull/101320

It makes mentally parsing the stalebot's action a matter of milliseconds.

In a few months, we can look again at the statistics if actionability has improved by it and account for if further action as proposed in this RFC can be considered as a nonzero-sum game (or even negative sum game as some people claimed).

blaggacao commented:
@Mic92 Is it possible to mark this RFC dormant due to "further data gathering"?

lheckemann commented:
Like #74, would you like to close this until you're ready to proceed?

7c6f434c commented:
Re: original topic: just have seen a comment today that stale bot existing is silly…

blaggacao commented:
@lheckemann Fully agree, if that's the process (which it seems to be).
